python/lsst/daf/butler/gen2convert/writer.py

```diff
@@ -0,0 +1,316 @@
+                translators[datasetTypeName] = Translator.makeMatching(camera=camera, datasetType=gen2dst,
+                                                                       skyMaps=gen2repo.skyMaps,
+                                                                       skyMapNames=skyMapNamesByCoaddName)
+            except NoSkyMapError:
+                log.warn("No SkyMap associated with DatasetType %s in %s; skipping.",
+                         datasetTypeName, gen2repo.root)
+                continue
+            if datasetTypeName in self.datasetTypes:
+                continue
+            storageClassName = scConfig.get(gen2dst.python, None)
+            if storageClassName is not None:
+                storageClass = scFactory.getStorageClass(storageClassName)
+            else:
+                log.debug("No StorageClass name configured for %s; trying persistable '%s'",
+                          gen2dst.name, gen2dst.persistable)
+                try:
+                    storageClass = scFactory.getStorageClass(gen2dst.persistable)
+                except KeyError:
+                    storageClass = None
+                if storageClass is None:
+                    unqualified = gen2dst.python.split(".")[-1]
+                    log.debug("No StorageClass name configured for %s; trying unqualified python type '%s'",
+                              gen2dst.name, unqualified)
+                    try:
+                        storageClass = scFactory.getStorageClass(unqualified)
+                    except KeyError:
+                        log.warn("No StorageClass found for %s; skipping.", gen2dst.name)
+                        continue
+            log.debug("Using StorageClass %s for %s", storageClass.name, gen2dst.name)
+            self.datasetTypes[datasetTypeName] = DatasetType(
+                name=datasetTypeName,
+                storageClass=storageClass,
+                dataUnits=translators[datasetTypeName].gen3units
+            )
+        converted = ConvertedRepo(gen2repo, camera=camera, run=run, translators=translators)
+        # Add parent repositories first, so self.repos is sorted topologically.
+        for parent in gen2repo.parents:
+            self._addConvertedRepoSorted(parent)
+        # Now we can finally add the current repo to self.repos.
+        self.repos[gen2repo.root] = converted
+        return converted
+
+    def run(self, registry, datastore):
+        """Main driver for ConversionWriter.
+
+        Runs all steps to create a Gen3 Repo, aside from Camera registration
+        (we merely check that the needed Cameras, Sensors, and PhysicalFilters
+        have already been registered).
+        """
+        self.checkCameras(registry)
+        self.insertSkyMaps(registry)
+        self.insertObservations(registry)
+        self.insertDatasetTypes(registry)
+        self.insertDatasets(registry, datastore)
+        # TODO: associate parent repo datasets with child repo Collections
```

Comment:
Is this still open? Seems reasonably essential.

Reply:
Yes, I'm afraid it is, and I remember it being hard but don't remember exactly why.  We can work around it with a custom call to `query` in the ci_hsc drivers for the conversion for now.